File: NeuralNetwork/NeuralNetwork.cs
using System.ComponentModel;
using static Neural_Network_and_AI.MathUtil;

/// <summary>
/// Author: Justin Farley
/// 
/// This was fun to build
/// </summary>
namespace Neural_Network_and_AI
{
    public class NeuralNetwork
    {
        public enum PredictionMethod{
            Linear,
            Softmax
        }
        public float learningRate = 0.01f;
        private int hiddenLayers;
        private int hiddenNodesPerLayer;
        private int inputSize;
        private int outputSize;
        private List<InputNode> inputNodes = new List<InputNode>();
        private List<OutputNode> outputNodes = new List<OutputNode>();
        private Dictionary<int, List<HiddenNode>> hiddenNodes = new Dictionary<int, List<HiddenNode>>();
        private List<Weight> weights = new List<Weight>();


        public NeuralNetwork(float learningRate, int inputSize, int outputSize, int hiddenLayers, int hiddenNodesPerLayer)
        {
            this.learningRate = learningRate;
            this.hiddenLayers = hiddenLayers;
            this.hiddenNodesPerLayer = hiddenNodesPerLayer;
            this.inputSize = inputSize;
            this.outputSize = outputSize;
        }

        private float GetWeightValue(Random random, int prevLayerCount)
        {
            float fanIn = prevLayerCount; // number of incoming connections
            float scale = MathF.Sqrt(2f / fanIn);
            float weightValue = ((float)random.NextDouble() * 2f - 1f) * scale;
            return weightValue;
        }

        public void NetworkInit()
        {
            Random random = new Random();        


            //make input nodes
            for(int i = 0; i < inputSize; i++)
            {
                InputNode input = new InputNode(0);
                inputNodes.Add(input);
            }

            //Create first layer of hidden nodes to connect to input nodes
            hiddenNodes[0] = new List<HiddenNode>();    
            for(int j = 0; j < hiddenNodesPerLayer; j++)
            {
                HiddenNode hiddenNode = new HiddenNode(GetWeightValue(random, inputNodes.Count));
                hiddenNodes[0].Add(hiddenNode);
            }

            //For every hidden node in the FIRST layer
            foreach(var hiddenNode in hiddenNodes[0])
            {
                foreach(var inputNode in inputNodes)
                {
                    Weight weight = new Weight(inputNode, hiddenNode, GetWeightValue(random, inputNodes.Count));
                    weights.Add(weight);
                }
            }

            //Starting from 1 to skip the initial layer we made already
            //Create these layers of hidden nodes
            for(int i = 1; i < hiddenLayers; i++)
            {
                hiddenNodes[i] = new List<HiddenNode>();
                for(int j = 0; j < hiddenNodesPerLayer; j++)
                {
                    HiddenNode hiddenNode = new HiddenNode(GetWeightValue(random, hiddenNodesPerLayer));
                    hiddenNodes[i].Add(hiddenNode);
                }
            }

            foreach(KeyValuePair<int, List<HiddenNode>> kvp in hiddenNodes)
            {
                foreach(var hiddenNode in kvp.Value)
                {
                    if(kvp.Key == 0) continue; //skip the first layer since we already connected it to the input nodes
                    foreach(var prevHiddenNode in hiddenNodes[kvp.Key - 1])
                    {
                        Weight weight = new Weight(prevHiddenNode, hiddenNode, GetWeightValue(random, hiddenNodesPerLayer));
                        weights.Add(weight);
                    }
                }
            }

            //Now create output nodes
            for(int i = 0; i < outputSize; i++)
            {
                OutputNode outputNode = new OutputNode((float)random.NextDouble());
                outputNodes.Add(outputNode);
                foreach(var hiddenNode in hiddenNodes[hiddenLayers - 1])
                {
                    Weight weight = new Weight(hiddenNode, outputNode, GetWeightValue(random, hiddenNodesPerLayer));
                    weights.Add(weight);
                }
            }
        }

        private List<float> ForwardPass(List<float> input, PredictionMethod method = PredictionMethod.Linear, List<float> expected = null)
        {

            for(int i = 0; i < inputNodes.Count; i++)
            {
                inputNodes[i].value = input[i];
            }

            foreach(var kvp in hiddenNodes)
            {
                foreach(var hiddenNode in kvp.Value)
                {
                    hiddenNode.Activation();
                }
            }

            List<float> predictedValues = new List<float>();

            for(int j = 0; j < outputNodes.Count; j++)
            {
                var outputNode = outputNodes[j];
                outputNode.Prediction();
                if(expected != null && method == PredictionMethod.Linear)
                    outputNode.error = GetOutputNodeError(outputNode.value, expected[j]);
                predictedValues.Add(outputNode.value);
            }
            if(method != PredictionMethod.Softmax)
                Console.WriteLine($"Loss (MSE): {MathUtil.MSELoss(expected ?? new List<float>(), predictedValues)}");

            if(method == PredictionMethod.Softmax)
            {
                predictedValues = new List<float>();
                var rawOutputs = outputNodes.Select(x => x.value).ToArray();

                float max = rawOutputs.Max();
                float sums = rawOutputs.Sum(x => MathF.Exp(x - max));

                for (int i = 0; i < outputNodes.Count; i++)
                {
                    float softmaxVal = MathF.Exp(rawOutputs[i] - max) / sums;
                    outputNodes[i].value = softmaxVal;
                    predictedValues.Add(softmaxVal);

                    if(expected != null)
                        outputNodes[i].error = softmaxVal - expected[i];
                }
            }
            return predictedValues;
        }

        public void BackwardsPass()
        {
            //START BACKWARDS PASS
                Dictionary<Weight, float> weightGradients = new Dictionary<Weight, float>(); 
                Dictionary<Node, float> biasGradients = new Dictionary<Node, float>(); 
                foreach(var outputNode in outputNodes)
                {
                    foreach(var weight in outputNode.incomingWeights)
                    {
                        //SOLVE OUTPUT WEIGHT GRADIENTS
                        float gradient = GetOutputParameterGradient(outputNode.error, weight.prev.value);
                        weightGradients[weight] = gradient;
                    }
                    //SOLVE OUTPUT BIAS GRADIENTS
                    biasGradients[outputNode] = GetOutputParameterGradient(outputNode.error, 1f); //bias is always 1
                }

                //UPDATE HIDDEN ERRORS AND RELU DERIVATIVES
                for (int j = hiddenLayers - 1; j >= 0; j--)
                {
                    foreach (var hiddenNode in hiddenNodes[j])
                    {
                        float error = 0;
                        foreach (var outgoingWeight in hiddenNode.outgoingWeights)
                        {
                            error += outgoingWeight.next.error * outgoingWeight.value;
                        }
                        hiddenNode.error = error * (hiddenNode.value > 0 ? 1.0f : 0.0f); 
                    }
                }
                // SOLVE HIDDEN LAYER(S) WEIGHT GRADIENTS AND INPUT LAYER WEIGHT GRADIENTS
                for(int j = hiddenLayers - 1; j >= 1; j--)
                {
                    foreach(var hiddenNode in hiddenNodes[j])
                    {
                        foreach(var weight in hiddenNode.incomingWeights)
                        {
                            float gradient = hiddenNode.error * weight.prev.value;;
                            weightGradients[weight] = gradient;
                        }
                        biasGradients[hiddenNode] = hiddenNode.error; //bias is always 1
                    }
                }

                //SOLVE INPUT LAYER WEIGHT GRADIENTS
                foreach(var hiddenNode in hiddenNodes[0])
                {
                    foreach(var weight in hiddenNode.incomingWeights)
                    {
                        float gradient = hiddenNode.error * weight.prev.value;
                        weightGradients[weight] = gradient;
                    }
                    biasGradients[hiddenNode] = hiddenNode.error; //bias is always 1
                }

                //UPDATE PARAMETERS!!
                foreach(var weight in weights)
                {
                    weight.value = UpdateParameter(weight.value, learningRate, weightGradients[weight]);
                }

                //Console.WriteLine($"First weight: {inputNodes[0].outgoingWeights[0].value}");

                foreach(var kvp in hiddenNodes)
                {
                    foreach(var hiddenNode in kvp.Value)
                    {
                        hiddenNode.bias = UpdateParameter(hiddenNode.bias, learningRate, biasGradients[hiddenNode]);
                    }
                }   
                
                // Apply output node bias updates (if computed)
                foreach (var outputNode in outputNodes)
                {
                    outputNode.bias = UpdateParameter(outputNode.bias, learningRate, biasGradients[outputNode]);
                }
                weightGradients.Clear();
                biasGradients.Clear();
        }

        public List<float> Predict(List<float> input, PredictionMethod method = PredictionMethod.Linear)
        {
            List<float> results = ForwardPass(input, method);
            return results;
        }

        public void Train(List<List<float>> inputs, List<List<float>> expected, int iterations, PredictionMethod method = PredictionMethod.Linear)
        {
            for(int j = 0; j < iterations; j++)
            {
                float totalLoss = 0;
                for(int i = 0; i < inputs.Count; i++)
                {
                    var preds = ForwardPass(inputs[i], method, expected[i]);

                    BackwardsPass();              
                    totalLoss += CrossEntropyLoss(preds, expected[i]);
                }
                float avgLoss = totalLoss / inputs.Count;
                Console.WriteLine($"Epoch {j + 1}/{iterations} - Avg Loss: {avgLoss}");
            }
        }

        public void Train(List<float> inputs, List<float> expected, int iterations, PredictionMethod method = PredictionMethod.Linear)
        {
            for (int i = 0; i < iterations; i++)
            {
                // Forward pass
                ForwardPass(inputs, method, expected);

                BackwardsPass();
            }
        }
    }
}

File: NeuralNetwork/Python/NeuralNetwork.py
import math
import random
import time
from enum import Enum
from InputNode import InputNode
from HiddenNode import HiddenNode
from OutputNode import OutputNode
from Weight import Weight
from MathUtil import MathUtil
"""
Author: Justin Farley

This was fun to build
"""

class NeuralNetwork:

    class PredictionMethod(Enum):
        Linear = 0
        Softmax = 1

    def __init__(self, learningRate, inputSize, outputSize, hiddenLayers, hiddenNodesPerLayer):
        self.learningRate = learningRate
        self.hiddenLayers = hiddenLayers
        self.hiddenNodesPerLayer = hiddenNodesPerLayer
        self.inputSize = inputSize
        self.outputSize = outputSize

        self.inputNodes = []
        self.outputNodes = []
        self.hiddenNodes = {}  # dict[int, list[HiddenNode]]
        self.weights = []

    def GetAllWeights(self):
        return self.weights

    def GetHiddenLayers(self):
        return self.hiddenNodes

    def GetOutputNodes(self):
        return self.outputNodes

    def GetInputNodes(self):
        return self.inputNodes

    def GetNodeIndex(self, node):
        index = 0

        for i in range(len(self.inputNodes)):
            if self.inputNodes[i] == node:
                return index
            index += 1

        for key in sorted(self.hiddenNodes.keys()):
            for hiddenNode in self.hiddenNodes[key]:
                if hiddenNode == node:
                    return index
                index += 1

        for i in range(len(self.outputNodes)):
            if self.outputNodes[i] == node:
                return index
            index += 1

        return -1

    @staticmethod
    def GetWeightValue(prevLayerCount):
        fanIn = prevLayerCount
        scale = math.sqrt(2.0 / fanIn)
        weightValue = (random.random() * 2.0 - 1.0) * scale
        return weightValue

    def NetworkInit(self):

        # Input nodes
        for i in range(self.inputSize):
            inputNode = InputNode(0)
            self.inputNodes.append(inputNode)

        # First hidden layer
        self.hiddenNodes[0] = []
        for j in range(self.hiddenNodesPerLayer):
            hiddenNode = HiddenNode(self.GetWeightValue(len(self.inputNodes)))
            self.hiddenNodes[0].append(hiddenNode)

        for hiddenNode in self.hiddenNodes[0]:
            for inputNode in self.inputNodes:
                weight = Weight(inputNode, hiddenNode,
                                self.GetWeightValue(len(self.inputNodes)))
                self.weights.append(weight)

        # Remaining hidden layers
        for i in range(1, self.hiddenLayers):
            self.hiddenNodes[i] = []
            for j in range(self.hiddenNodesPerLayer):
                hiddenNode = HiddenNode(self.GetWeightValue(self.hiddenNodesPerLayer))
                self.hiddenNodes[i].append(hiddenNode)

        for key, layer in self.hiddenNodes.items():
            for hiddenNode in layer:
                if key == 0:
                    continue
                for prevHiddenNode in self.hiddenNodes[key - 1]:
                    weight = Weight(prevHiddenNode, hiddenNode,
                                    self.GetWeightValue(self.hiddenNodesPerLayer))
                    self.weights.append(weight)

        # Output nodes
        for i in range(self.outputSize):
            outputNode = OutputNode(random.random())
            self.outputNodes.append(outputNode)

            for hiddenNode in self.hiddenNodes[self.hiddenLayers - 1]:
                weight = Weight(hiddenNode, outputNode,
                                self.GetWeightValue(self.hiddenNodesPerLayer))
                self.weights.append(weight)

    def ForwardPass(self, input, method=PredictionMethod.Linear, expected=None):

        for i in range(len(self.inputNodes)):
            self.inputNodes[i].value = input[i]

        for key in self.hiddenNodes:
            for hiddenNode in self.hiddenNodes[key]:
                hiddenNode.Activation()

        predictedValues = []

        for j in range(len(self.outputNodes)):
            outputNode = self.outputNodes[j]
            outputNode.Prediction()

            if expected is not None and method == self.PredictionMethod.Linear:
                outputNode.error = MathUtil.GetOutputNodeError(outputNode.value, expected[j])

            predictedValues.append(outputNode.value)

        if method != self.PredictionMethod.Softmax and expected is not None:
            print(f"Loss (MSE): {MathUtil.MSELoss(expected, predictedValues)}")

        if method == self.PredictionMethod.Softmax:

            predictedValues = []
            rawOutputs = [x.value for x in self.outputNodes]

            maxVal = max(rawOutputs)
            sums = sum(math.exp(x - maxVal) for x in rawOutputs)

            for i in range(len(self.outputNodes)):
                softmaxVal = math.exp(rawOutputs[i] - maxVal) / sums
                self.outputNodes[i].value = softmaxVal
                predictedValues.append(softmaxVal)

                if expected is not None:
                    self.outputNodes[i].error = softmaxVal - expected[i]

        return predictedValues

    def BackwardsPass(self):

        weightGradients = {}
        biasGradients = {}

        # Output gradients
        for outputNode in self.outputNodes:
            for weight in outputNode.incomingWeights:
                gradient = MathUtil.GetOutputParameterGradient(
                    outputNode.error, weight.prev.value)
                weightGradients[weight] = gradient

            biasGradients[outputNode] = MathUtil.GetOutputParameterGradient(
                outputNode.error, 1.0)

        # Hidden error calculation
        for j in range(self.hiddenLayers - 1, -1, -1):
            for hiddenNode in self.hiddenNodes[j]:
                error = 0
                for outgoingWeight in hiddenNode.outgoingWeights:
                    error += outgoingWeight.next.error * outgoingWeight.value

                hiddenNode.error = error * (1.0 if hiddenNode.value > 0 else 0.0)

        # Hidden layer gradients
        for j in range(self.hiddenLayers - 1, 0, -1):
            for hiddenNode in self.hiddenNodes[j]:
                for weight in hiddenNode.incomingWeights:
                    gradient = hiddenNode.error * weight.prev.value
                    weightGradients[weight] = gradient

                biasGradients[hiddenNode] = hiddenNode.error

        # Input layer gradients
        for hiddenNode in self.hiddenNodes[0]:
            for weight in hiddenNode.incomingWeights:
                gradient = hiddenNode.error * weight.prev.value
                weightGradients[weight] = gradient

            biasGradients[hiddenNode] = hiddenNode.error

        # Update weights
        for weight in self.weights:
            weight.value = MathUtil.UpdateParameter(
                weight.value,
                self.learningRate,
                weightGradients[weight]
            )

        # Update hidden biases
        for key in self.hiddenNodes:
            for hiddenNode in self.hiddenNodes[key]:
                hiddenNode.bias = MathUtil.UpdateParameter(
                    hiddenNode.bias,
                    self.learningRate,
                    biasGradients[hiddenNode]
                )

        # Update output biases
        for outputNode in self.outputNodes:
            outputNode.bias = MathUtil.UpdateParameter(
                outputNode.bias,
                self.learningRate,
                biasGradients[outputNode]
            )

        weightGradients.clear()
        biasGradients.clear()

    def Predict(self, input, method=PredictionMethod.Linear):
        return self.ForwardPass(input, method)

    def Train(self, inputs, expected, iterations,
              method=PredictionMethod.Linear):

        print(f"Began Training @ {time.strftime('%H:%M:%S')}")

        for j in range(iterations):

            start = time.time()
            totalLoss = 0

            for i in range(len(inputs)):
                preds = self.ForwardPass(inputs[i], method, expected[i])
                self.BackwardsPass()
                totalLoss += MathUtil.CrossEntropyLoss(preds, expected[i])

            avgLoss = totalLoss / len(inputs)
            elapsed = int(time.time() - start)

            print(f"Epoch {j+1}/{iterations} - Avg Loss: {avgLoss}, "
                  f"Time for this epoch: {elapsed//60}m and {elapsed%60}s "
                  f"@ {time.strftime('%H:%M:%S')}")

    def TrainSingle(self, inputs, expected, iterations,
                    method=PredictionMethod.Linear):

        for i in range(iterations):
            self.ForwardPass(inputs, method, expected)
            self.BackwardsPass()
